### PR TITLE
winapi: Implement SetEndOfFile()

### LINKED ISSUE #471: `SetEndOfFile` is missing
Needed by a couple of projects, and looks easy to implement, so leaving it here, as a potential task.

### PR DESCRIPTION
**Pull request by kosmas12**
Closes #471

Please, excuse the failed Ubuntu CI on my repo - It's because my fork is really old.

### CODE REVIEW

File: lib/winapi/fileio.c [hunk 1]
```diff
@@ -287,6 +287,32 @@ BOOL SetFilePointerEx (HANDLE hFile, LARGE_INTEGER liDistanceToMove, PLARGE_INTE
...
+    fullCurrentPos = fullCurrentPos << 32 | lowPosDWORD;
+
+
+    IO_STATUS_BLOCK ioStatusBlock;
+    FILE_END_OF_FILE_INFORMATION fileEOFInfo;
+    fileEOFInfo.EndOfFile.QuadPart = fullCurrentPos;
+
+    NTSTATUS status = NtSetInformationFile(hFile, &ioStatusBlock, &fileEOFInfo, sizeof(fileEOFInfo), FileEndOfFileInformation);
+    if (NT_SUCCESS(status)) {
+        return FALSE;
```
Comment (thrimbor) [inline lib/winapi/fileio.c:309]: lastError needs to get set here, you can check other functions to see how.

File: lib/winapi/fileio.c [hunk 2]
```diff
@@ -287,6 +287,32 @@ BOOL SetFilePointerEx (HANDLE hFile, LARGE_INTEGER liDistanceToMove, PLARGE_INTE
     }
 }
 
+BOOL SetEndOfFile (HANDLE hFile)
+{
+    // Get current position in file
+    LONG highPosDWORD = 0;
+    PLONG lpHighPosDWORD = &highPosDWORD;
+    DWORD lowPosDWORD = SetFilePointer(hFile, 0, lpHighPosDWORD, FILE_CURRENT);
```
Comment (thrimbor) [inline lib/winapi/fileio.c:295]: `SetFilePointer` is not an ideal choice as its handling of 64 bit offsets is awkward. `SetFilePointerEx` is better, but I'd prefer it this code used `NtQueryInformationFile` directly.

File: lib/winapi/fileio.c [hunk 3]
```diff
@@ -287,6 +287,32 @@ BOOL SetFilePointerEx (HANDLE hFile, LARGE_INTEGER liDistanceToMove, PLARGE_INTE
...
+    FILE_END_OF_FILE_INFORMATION fileEOFInfo;
+    fileEOFInfo.EndOfFile.QuadPart = fullCurrentPos;
+
+    NTSTATUS status = NtSetInformationFile(hFile, &ioStatusBlock, &fileEOFInfo, sizeof(fileEOFInfo), FileEndOfFileInformation);
+    if (NT_SUCCESS(status)) {
+        return FALSE;
+    }
+    else {
+        return TRUE;
+    }
```
Comment (thrimbor) [inline lib/winapi/fileio.c:313]: That's awkward formatting. Personally I prefer to keep the `return TRUE;` case outside of the if.

File: lib/winapi/fileio.c [hunk 4]
```diff
@@ -287,6 +287,32 @@ BOOL SetFilePointerEx (HANDLE hFile, LARGE_INTEGER liDistanceToMove, PLARGE_INTE
...
+}
```
Comment (thrimbor) [inline lib/winapi/fileio.c:314]: `FileAllocationInformation` needs treatment equivalent to `FileEndOfFileInformation` here.

**Review by thrimbor — CHANGES_REQUESTED** [review]

File: lib/winapi/fileio.c (hunk 1 above)
Comment (kosmas12) [reply lib/winapi/fileio.c:309]: I was expecting this, but problem is, I don't know what to set the error to. Is there some kind of standard way we should follow to mimic Windows? Or can we just have our own error messages? Still not sure about the proper way to contribute to WinAPI code.

File: lib/winapi/fileio.c (hunk 2 above)
Comment (kosmas12) [reply lib/winapi/fileio.c:295]: I'll check that possibility out.

File: lib/winapi/fileio.c [hunk 5]
```diff
@@ -287,6 +287,32 @@ BOOL SetFilePointerEx (HANDLE hFile, LARGE_INTEGER liDistanceToMove, PLARGE_INTE
...
+{
+    // Get current position in file
+    LONG highPosDWORD = 0;
+    PLONG lpHighPosDWORD = &highPosDWORD;
+    DWORD lowPosDWORD = SetFilePointer(hFile, 0, lpHighPosDWORD, FILE_CURRENT);
+
+    // Combine low and high position DWORDs for full position
+    LONGLONG fullCurrentPos;
+    fullCurrentPos = highPosDWORD;
+    fullCurrentPos = fullCurrentPos << 32 | lowPosDWORD;
```
Comment (JayFoxRox) [inline lib/winapi/fileio.c:300]: Not sure if `<< 32` will work properly. You might have to use `<< 32LL`. Please verify

Also check operator precedence please.

File: lib/winapi/fileio.c (hunk 1 above)
Comment (JayFoxRox) [reply lib/winapi/fileio.c:309]: *(for future reviews: If you are aware of missing code / existing issues, please leave a comment in the code or, better yet, add a review comment yourself)*

File: lib/winapi/fileio.c (hunk 3 above)
Comment (JayFoxRox) [reply lib/winapi/fileio.c:313]: It's not just formatting, it's also required for some IDEs (or even compilers?)

**Comment by thrimbor** [comment]
Closing this as https://github.com/XboxDev/nxdk/pull/583 has been merged

**Pull request closed without merging**